### PR TITLE
Add `--preferred-encoding (gzip|brotli)` to use when tile is not pre-encoded by source

### LINKED ISSUE #1178: Implement `--preferred-encoding (gzip|brotli)` configuration
Brotli encoding produces smaller content, but is slower to encode.  GZip is the opposite of that. Until v0.11, Martin used GZIP as the default, but later switched to Brotli, causing slower performance for some users, as reported in https://github.com/maplibre/martin/issues/1172

Per https://github.com/maplibre/martin/issues/1172#issuecomment-1926047552, we need a new configuration parameter in both CLI and in config.  CLI could use `--preferred-encoding (gzip|brotli)` (should also allow `"br"` for consistency with `Accept-Encoding` header, and config:

```
# root level
preferred_encoding: brotli
```

This setting will only work if the client provided [Accept-Encoding header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept-Encoding) where `gzip` and `br` have the same top `q` value (the `q` value could be missing for both, which also makes it the same 1.0).  In theory, the client may prefer `identity`, which we also support.

### PR DESCRIPTION
**Pull request by sharkAndshark**
Try to fix #1178 

- [x] Add optional `--preferred-encoding` to cli, `br` `brotli` and `gzip` are allowed
- [x] Add optional `preferred-encoding` to config, `br` `brotli` and `gzip` are allowed
- [x] Add test
- [x] Update doc

### CODE REVIEW

File: martin/src/args/srv.rs [hunk 1]
```diff
@@ -10,6 +12,8 @@ pub struct SrvArgs {
     /// Number of web server workers
     #[arg(short = 'W', long)]
     pub workers: Option<usize>,
+    #[arg(help = "to do", short, long)]
+    pub preferred_encoding: Option<String>,
```
Comment (nyurik) [inline martin/src/args/srv.rs]: Don't use String here. Create a new `enum PreferredEncoding` similar to `BoundsCalcType`.  Note that serde supports aliases (i.e. support both br and brotli in one value).

File: martin/src/args/srv.rs [hunk 2]
```diff
@@ -10,6 +12,8 @@ pub struct SrvArgs {
     /// Number of web server workers
     #[arg(short = 'W', long)]
     pub workers: Option<usize>,
+    #[arg(help = "to do", short, long)]
```
Comment (nyurik) [inline martin/src/args/srv.rs]: don't use `help`, use the `/// ...` right above this line

File: martin/src/srv/server.rs [hunk 3]
```diff
@@ -115,7 +115,8 @@ pub fn new_server(config: SrvConfig, state: ServerState) -> MartinResult<(Server
 
         let app = App::new()
             .app_data(Data::new(state.tiles.clone()))
-            .app_data(Data::new(state.cache.clone()));
+            .app_data(Data::new(state.cache.clone()))
+            .app_data(Data::new(state.preferred_encoding));
```
Comment (nyurik) [inline martin/src/srv/server.rs]: I think we should store `SrvConfig` object in the data.  Read the steps in https://github.com/maplibre/martin/issues/1185 -- here you have almost the same thing - a configuration value you can use in `get_tile` in `tiles.rs`

**Review by nyurik — COMMENTED** [review]
good first step, left a few comments

File: martin/src/srv/tiles.rs [hunk 4]
```diff
@@ -184,9 +185,30 @@ impl<'a> DynTileSource<'a> {
                 }
             }
             if tile.info.encoding == Encoding::Uncompressed {
+                let ordered_encodings = if let Some(prefered) = self.prefered_encodings {
+                    match prefered {
+                        PreferredEncoding::Brotli => [
```
Comment (nyurik) [inline martin/src/srv/tiles.rs]: I am not too certain if the compiler can optimize these cases on the fly - maybe replace these with 2 statics, like we used to have with `SUPPORTED_ENCODINGS`:  

```rust
static PREFER_BROTLI_ENC: &[HeaderEnc] = &[ ... ];
static PREFER_GZIP_ENC: &[HeaderEnc] = &[ ... ];
```

and in this match statement just do this:

```rust
let ordered_encodings = match self.prefered_encodings {
  Some(PreferredEncoding::Gzip) => PREFER_GZIP_ENC,
  Some(PreferredEncoding::Brotli) | None => PREFER_BROTLI_ENC
};
```

**Review by nyurik — COMMENTED** [review]
love it!  One small nit, and seems it is pretty good to go!

File: martin/src/srv/tiles.rs [hunk 5]
```diff
@@ -93,7 +93,8 @@ impl<'a> DynTileSource<'a> {
             info,
             query_str,
             query_obj,
-            encodings,
+            accept_encodings: accpect_encodings,
```
Comment (nyurik) [inline martin/src/srv/tiles.rs]: maybe better rename these into `accept_enc` and `preferred_enc` (notice you misspelled the var there... and you are not the first - similar fate with [Referer header](https://en.wikipedia.org/wiki/HTTP_referer)) :)

One uncertainty - is this a LIST of accepted encodings, or the encoding that has been accepted (chosen)?  The name is a bit confusing - might need to make it clearer.

thx!

Comment (sharkAndshark) [reply martin/src/srv/tiles.rs]: I need to improve my English more XD 😂
Any better renaming suggestion? Like `accept_enc` => `browser_accept_enc`? I'm not sure...

File: martin/src/srv/tiles.rs (hunk 4 above)
Comment (sharkAndshark) [reply martin/src/srv/tiles.rs]: Thx!  Much better and efficiency.

File: martin/src/srv/tiles.rs (hunk 5 above)
Comment (sharkAndshark) [reply martin/src/srv/tiles.rs]: Or `browser_wanted_enc`?

File: docs/src/config-file.md [hunk 6]
```diff
@@ -27,6 +27,9 @@ worker_processes: 8
 # Amount of memory (in MB) to use for caching tiles [default: 512, 0 to disable]
 cache_size_mb: 1024
 
+# Preferred tiles encoding, gzip or brotli, default brotili. You could also use br as a shortcut for brotli
```
Comment (nyurik) [inline docs/src/config-file.md]:
```suggestion
# If the client accepts multiple compression formats, and the tile source is not pre-compressed, which compression should be used. `gzip` is faster, but `brotli` is smaller, and may be faster with caching.  Defaults to brotli.
```

File: docs/src/run-with-cli.md [hunk 7]
```diff
@@ -31,6 +31,11 @@ Options:
   -W, --workers <WORKERS>
           Number of web server workers
 
+      --preferred-encoding <PREFERRED_ENCODING>
```
Comment (nyurik) [inline docs/src/run-with-cli.md:34]: similar as above?

File: martin/src/args/srv.rs [hunk 8]
```diff
@@ -10,6 +12,19 @@ pub struct SrvArgs {
     /// Number of web server workers
     #[arg(short = 'W', long)]
     pub workers: Option<usize>,
+    /// Preferred tiles encoding. gzip or brotli, default brotili. You could also use br as a shortcut for brotli
```
Comment (nyurik) [inline martin/src/args/srv.rs]: same

File: martin/tests/mb_server_test.rs [hunk 9]
```diff
@@ -24,6 +25,12 @@ macro_rules! create_app {
                 ))
                 .app_data(actix_web::web::Data::new(::martin::NO_MAIN_CACHE))
                 .app_data(actix_web::web::Data::new(state.tiles))
+                .app_data(actix_web::web::Data::new(SrvConfig {
```
Comment (nyurik) [inline martin/tests/mb_server_test.rs]: for tests might as well use `SrvConfig::default()` here and below

**Review by nyurik — COMMENTED** [review]
nah, not worth it, will fix naming if needed later, good enough.  I made a few more minor suggestions

File: martin/src/srv/tiles.rs [hunk 10]
```diff
@@ -183,10 +196,17 @@ impl<'a> DynTileSource<'a> {
...
             if tile.info.encoding == Encoding::Uncompressed {
+                let ordered_encodings = match self.preferred_enc {
+                    Some(PreferredEncoding::Gzip) => PREFER_GZIP_ENC,
+                    Some(PreferredEncoding::Brotli) | None => PREFER_BROTLI_ENC,
+                };
+
                 // only apply compression if the content supports it
                 if let Some(HeaderEnc::Known(enc)) =
-                    accept_enc.negotiate(SUPPORTED_ENCODINGS.iter())
+                    // accept_enc.negotiate(SUPPORTED_ENCODINGS.iter())
```
Comment (nyurik) [inline martin/src/srv/tiles.rs]:
```suggestion
```

**Review by nyurik — APPROVED** [review]
I'm not too happy with the description, but can't think think too clearly. Will try in the morning. In the mean time, there is one line that i think was left by an accident, but good to go otherwise, so remove it and feel free to merge :)

File: martin/src/srv/tiles.rs (hunk 10 above)
Comment (sharkAndshark) [reply martin/src/srv/tiles.rs]: Oops. Deleted.

Comment (sharkAndshark) [reply martin/src/srv/tiles.rs]: "Preferred tile encoding when martin is negotiating with browser, only works if the tile itself is not pre-compressed. The default Brotli is much smaller(and may be faster with caching) but a little bit slower than gzip."
Well this version maybe much worse..

**Pull request merged**